File: hse/BPM161/lupuleac/2/LockFreeSet/src/main/java/ru/hse/lupuleac/lockfree/LockFreeSet.java
package ru.hse.lupuleac.lockfree;


import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.concurrent.atomic.AtomicMarkableReference;
import java.util.stream.Collectors;

public class LockFreeSet<T extends Comparable<T>> implements
        Set<T> {
    private Node head;

    public LockFreeSet() {
        head = new Node(null);
    }

    @Override
    public boolean add(T value) {
        while (true) {
            FindResult findResult = find(value);
            if (findResult.cur != null && findResult.cur.value
                    .compareTo(value) == 0) {
                return false;
            }
            Node node = new Node(value);
            node.next.set(findResult.cur, false);
            if (findResult.prev.next.compareAndSet(findResult.cur,
                    node, false, false)) {
                return true;
            }
        }
    }

    @Override
    public boolean remove(T value) {
        while (true) {
            FindResult findResult = find(value);
            Node nodeToBeRemoved = findResult.cur;

            if (nodeToBeRemoved == null || nodeToBeRemoved.value
                    .compareTo(value) != 0) {
                return false;
            }
            Node next = nodeToBeRemoved.nextNode();
            if (!nodeToBeRemoved.next.attemptMark(next, true)) {
                continue;
            }

            if (findResult.prev.next.compareAndSet(nodeToBeRemoved,
                    next, false, false)) {
                return true;
            }
        }
    }


    private FindResult find(T key) {
        Node prev = head;
        Node cur = head.nextNode();
        while (cur != null && cur.value.compareTo(key) <
                0) {
            prev = cur;
            cur = cur.nextNode();
        }
        return new FindResult(prev, cur);
    }

    @Override
    public boolean contains(T value) {
        Node cur = head.nextNode();
        while (cur != null && (cur.value.compareTo
                        (value) < 0)) {
            cur = cur.nextNode();
        }
        return cur != null && cur.value.compareTo(value) == 0 && cur
                .exists();
    }

    @Override
    public boolean isEmpty() {
        Node cur = head.nextNode();
        while (cur != null) {
            if (cur.exists()) {
                return false;
            }
            cur = cur.nextNode();
        }
        return true;
    }

    private List<Node> getNodes() {
        Node cur = head.nextNode();
        List<Node> res = new ArrayList<>();
        while (cur != null) {
           if (cur.exists()){
               res.add(cur);
           }
           cur = cur.nextNode();
        }
        return res;
    }

    public List<T> scan() {
        List<Node> snapshot = getNodes();
        while (true) {
            List<Node> secondSnapshot = getNodes();

            if (snapshot.equals(secondSnapshot)) {
                return secondSnapshot.stream()
                        .map(t -> t.value)
                        .collect(
                                Collectors.toList()
                        );
            }
            snapshot = secondSnapshot;
        }
    }


    @Override
    public Iterator<T> iterator() {
        return scan().iterator();
    }


    private class Node {
        private Node(T value) {
            this.value = value;
            next = new AtomicMarkableReference<>(null, false);
        }

        private T value;
        private AtomicMarkableReference<Node> next;

        public boolean exists() {
            return !next.isMarked();
        }

        public Node nextNode() {
            return next.getReference();
        }
    }

    private class FindResult {
        private Node prev;
        private Node cur;

        private FindResult(Node prev, Node cur) {
            this.prev = prev;
            this.cur = cur;
        }
    }
}

File: hse/BPM161/kazakov/2/src/main/kotlin/ru/hse/spb/kazakov/LockFreeSet.kt
package ru.hse.spb.kazakov

import java.lang.IllegalStateException
import java.util.concurrent.atomic.AtomicMarkableReference


/**
 * Lock-Free множество.
 * @param <T> Тип ключей
 */
class LockFreeSet<T : Comparable<T>> : Iterable<T> {
    private val tail = BoundNode<T>()
    private val head = BoundNode(tail)

    /**
     * Lock-free добавление ключа к множеству.
     *
     * @param value значение ключа
     * @return false если value уже существует в множестве, true если элемент был добавлен
     */
    fun add(value: T): Boolean {
        while (true) {
            val (lower, ceiling) = find(value)
            if (ceiling != tail && ceiling.value == value) {
                return false
            }
            if (lower.casNext(ceiling, ValueNode(value, ceiling))) {
                return true
            }
        }
    }

    /**
     * Lock-free удаление ключа из множества.
     *
     * @param value значение ключа
     * @return false если ключ не был найден, true если ключ успешно удален
     */
    fun remove(value: T): Boolean {
        while (true) {
            val (lower, ceiling) = find(value)
            if (ceiling == tail || ceiling.value != value) {
                return false
            }
            val next = ceiling.next
            if (ceiling.casRemoveMark(next)) {
                lower.casNext(ceiling, next)
                return true
            }
        }
    }

    /**
     * Wait-free проверка наличия ключа в множестве.
     *
     * @param value значение ключа
     * @return true если элемент содержится в множестве, иначе - false
     */
    fun contains(value: T): Boolean {
        var current: Node<T> = head.next

        while (current != tail) {
            if (current.value == value && !current.isRemoved) {
                return true
            }
            current = current.next
        }

        return false
    }

    /**
     * Lock-free проверка множества на пустоту.
     *
     * @return true если множество пусто, иначе - false
     */
    fun isEmpty() = makeAtomicSnapshot().isEmpty()

    /**
     * Возвращает lock-free итератор для множества.
     */
    override fun iterator(): Iterator<T> = makeAtomicSnapshot().iterator()

    private fun find(value: T): Pair<Node<T>, Node<T>> {
        while (true) {
            var lower: Node<T> = head
            var ceiling = head.next
            var lowerNext = ceiling

            while (ceiling != tail && (ceiling.isRemoved || ceiling.value < value)) {
                if (!ceiling.isRemoved) {
                    lower = ceiling
                    lowerNext = ceiling.next
                }
                ceiling = ceiling.next
            }

            if (lower.casNext(lowerNext, ceiling)) {
                if (ceiling.isRemoved) {
                    continue
                }
                return Pair(lower, ceiling)
            }
        }
    }

    private fun makeAtomicSnapshot(): List<T> {
        var firstSnapshot = makeSnapshot()
        var secondSnapshot = makeSnapshot()

        while (firstSnapshot != secondSnapshot) {
            firstSnapshot = makeSnapshot()
            secondSnapshot = makeSnapshot()
        }

        return firstSnapshot.map { it.value }
    }

    private fun makeSnapshot(): List<Node<T>> {
        var current: Node<T> = head.next
        val values = mutableListOf<Node<T>>()

        while (current != tail) {
            if (!current.isRemoved) {
                values.add(current)
            }
            current = current.next
        }

        return values
    }

    private abstract class Node<T : Comparable<T>> {
        private val nextRef: AtomicMarkableReference<Node<T>>

        constructor() {
            nextRef = AtomicMarkableReference(this, false)
        }

        constructor(next: Node<T>) {
            nextRef = AtomicMarkableReference(next, false)
        }

        val next: Node<T>
            get() = nextRef.reference

        val isRemoved: Boolean
            get() = nextRef.isMarked

        abstract val value: T

        fun casNext(currentNext: Node<T>, newNext: Node<T>) = nextRef.compareAndSet(currentNext, newNext, false, false)

        fun casRemoveMark(currentNext: Node<T>) = nextRef.compareAndSet(currentNext, currentNext, false, true)
    }

    private class ValueNode<T : Comparable<T>>(override val value: T, next: Node<T>) : Node<T>(next)

    private class BoundNode<T : Comparable<T>> : Node<T> {
        constructor()

        constructor(next: Node<T>) : super(next)

        override val value: T
            get() = throw IllegalStateException("Bound node value queried")
    }
}